libraries/account-resolution/src/state.rs

```diff
@@ -0,0 +1,447 @@
+impl ExtraAccountMetas {
+    /// Initialize pod slice data for the given instruction and account infos
+    /// into the given buffer
+    pub fn init<'a, T: TlvDiscriminator, M: 'a>(
+        data: &mut [u8],
+        account_infos: &'a [M],
+    ) -> Result<(), ProgramError>
+    where
+        PodAccountMeta: From<&'a M>,
+    {
+        let mut state = TlvStateMut::unpack(data).unwrap();
+        let tlv_size = PodSlice::<PodAccountMeta>::size_of(account_infos.len())?;
+        let bytes = state.alloc::<T>(tlv_size)?;
+        let mut extra_account_metas = PodSliceMut::init(bytes)?;
+        for account_info in account_infos {
+            extra_account_metas.push(PodAccountMeta::from(account_info))?;
+        }
+        Ok(())
+    }
+
+    /// Initialize a TLV entry for the given discriminator, populating the data
+    /// with the given account infos
+    pub fn init_with_account_infos<T: TlvDiscriminator>(
+        data: &mut [u8],
+        account_infos: &[AccountInfo<'_>],
+    ) -> Result<(), ProgramError> {
+        Self::init::<T, AccountInfo>(data, account_infos)
+    }
+
+    /// Initialize a TLV entry for the given discriminator, populating the data
+    /// with the given account metas
+    pub fn init_with_account_metas<T: TlvDiscriminator>(
+        data: &mut [u8],
+        account_metas: &[AccountMeta],
+    ) -> Result<(), ProgramError> {
+        Self::init::<T, AccountMeta>(data, account_metas)
+    }
+
+    /// Get the byte size required to hold `num_items` items
+    pub fn size_of(num_items: usize) -> Result<usize, ProgramError> {
+        Ok(TlvStateBorrowed::get_base_len()
+            .saturating_add(PodSlice::<PodAccountMeta>::size_of(num_items)?))
+    }
+
+    /// Add the additional account metas to an existing instruction
+    pub fn add_to_instruction<T: TlvDiscriminator>(
+        instruction: &mut Instruction,
+        data: &[u8],
+    ) -> Result<(), ProgramError> {
+        let state = TlvStateBorrowed::unpack(data)?;
+        let bytes = state.get_bytes::<T>()?;
+        let extra_account_metas = PodSlice::<PodAccountMeta>::unpack(bytes)?;
+        instruction
+            .accounts
+            .extend(extra_account_metas.data().iter().map(|m| m.into()));
+        Ok(())
+    }
+
+    /// Add the additional account metas and account infos for a CPI
+    pub fn add_to_cpi_instruction<'a, T: TlvDiscriminator>(
+        cpi_instruction: &mut Instruction,
+        cpi_account_infos: &mut Vec<AccountInfo<'a>>,
+        data: &[u8],
+        account_infos: &[AccountInfo<'a>],
+    ) -> Result<(), ProgramError> {
+        let state = TlvStateBorrowed::unpack(data)?;
+        let bytes = state.get_bytes::<T>()?;
+        let extra_account_metas = PodSlice::<PodAccountMeta>::unpack(bytes)?;
+        cpi_instruction
+            .accounts
+            .extend(extra_account_metas.data().iter().map(Into::into));
+
+        for account_meta in extra_account_metas.data().iter().map(AccountMeta::from) {
```

Comment:
It's a good point... I was aiming to avoid the additional allocation required for the map. Maybe we could eventually expose another version that uses a map internally? That way you can go with "allocation-free-but-maybe-loads-of-compute-units" or "allocs-but-fewer-compute-units". What do you think?

Reply:
Oh of course, we can get this down to one conversion by pushing into the vector, ie:

```
        for account_meta in extra_account_metas.data().iter().map(AccountMeta::from) {
            let account_info = account_infos
                .iter()
                .find(|&x| *x.key == account_meta.pubkey)
                .ok_or(AccountResolutionError::IncorrectAccount)?
                .clone();
            cpi_account_infos.push(account_info);
            cpi_instruction.accounts.push(account_meta);
        }
```

Thanks!